src/mds/Server.cc

```diff
@@ -5089,6 +5126,64 @@ void Server::handle_client_setattr(MDRequestRef& mdr)
       cur->add_waiter(CInode::WAIT_TRUNC, new C_MDS_RetryRequest(mdcache, mdr));
       return;
     }
+
+    if (truncating_smaller && req->get_data().length()) {
+      struct ceph_fscrypt_last_block_header header;
+      memset(&header, 0, sizeof(header));
+      auto bl = req->get_data().cbegin();
+      bufferlist *data = new bufferlist();
+      DECODE_START(1, bl);
+      decode(header.objver, bl);
+      decode(header.file_offset, bl);
+      decode(header.block_size, bl);
+      DECODE_FINISH(bl);
+
+      dout(20) << __func__ << " mdr->objver:" << mdr->objver
+               << " mdr->retry:" << mdr->retry
+               << " header.objver: " << header.objver
+               << " header.file_offset: " << header.file_offset
+               << " header.block_size: " << header.block_size
+               << dendl;
+
+      if (!header.objver) {
+        dout(20) << __func__ << " The last block is located in a file hole!"
+		 << dendl;
+      } else if (mdr->objver) {
+        if (mdr->objver != header.objver) {
```

Comment:
This is not at all what I had in mind. I had assumed that the objver you got from the client would be used to build an assert_ver op that preceded the truncate and write of the backing object. Here though, you're comparing the objver in the setattr request to mdr->objver.

Where does mdr->objver come from and how do you know that it's the same value as the backing object of the last block? I guess you're relying on Fx ensuring that no changes can be made? That may be OK then.

Reply:
The `mdr->objver`, which initially is 0, is read from the Rados in Line#5183 below. Yeah, since the setattr has `xlock(filelock)` at the beginning, so no client could write to the file during this time.

We need to make sure the `header.objver` will be the latest after xlocked the filelock as you mentioned above.

Reply:
We need to verify the object versions before submitting the mdlog for it, or it will be more complicated if we need to let client retry if they won't match.

Reply:
Ok, I suppose we can live with that. It's an extra round trip between the MDS and OSD, but that's not the end of the world. I'll mark this resolved.